### PR TITLE
raft: don't allocate slice and sort on every commit

### PR DESCRIPTION
**Pull request by nvanbenschoten**

### CODE REVIEW

**Comment by xiang90** [comment]
will this bring in an observable perf improvement for cockroachdb?

**Comment by nvanbenschoten** [comment]
I doubt it will be observable, but saving 2 allocations (the slice and the `sort.reverse`) and often avoiding the need to `Sort` for each committed entry seems like a win.

**Comment by nvanbenschoten** [comment]
Here's a `benchstat` diff of this change running on `Proposal3Nodes`:
```
name              old time/op    new time/op    delta
Proposal3Nodes-4    5.04µs ±26%    4.62µs ±32%   -8.35%  (p=0.033 n=20+20)

name              old alloc/op   new alloc/op   delta
Proposal3Nodes-4    4.58kB ±42%    4.34kB ±69%     ~     (p=0.192 n=20+20)

name              old allocs/op  new allocs/op  delta
Proposal3Nodes-4      11.5 ±39%       9.7 ±34%  -15.65%  (p=0.013 n=20+20)
```

The change is also addressing a TODO, so we should remove that TODO if we don't want to optimize this.

**Comment by xiang90** [comment]
@nvanbenschoten can you rebase?

**Comment by xiang90** [comment]
@siddontang can you take a look?

**Comment by nvanbenschoten** [comment]
@xiang90 I just rebased this off master.

**Comment by siddontang** [comment]
@xiang90 ok

**Comment by siddontang** [comment]
I suggest using a new PR for the change of Raft learner. 

Rest LGTM

**Comment by nvanbenschoten** [comment]
@siddontang TFTR. I pulled the second commit out into https://github.com/coreos/etcd/pull/9088.

File: raft/raft.go [hunk 1]
```diff
@@ -547,13 +548,23 @@ func (r *raft) bcastHeartbeatWithCtx(ctx []byte) {
 // the commit index changed (in which case the caller should call
 // r.bcastAppend).
 func (r *raft) maybeCommit() bool {
-	// TODO(bmizerany): optimize.. Currently naive
-	mis := make(uint64Slice, 0, len(r.prs))
+	if cap(r.matchBuf) < len(r.prs) {
```
Comment (xiang90) [inline raft/raft.go]: probably document this optimization for avoiding allocations on slice creating and sorting.

File: raft/raft.go [hunk 2]
```diff
@@ -547,13 +548,23 @@ func (r *raft) bcastHeartbeatWithCtx(ctx []byte) {
...
+		r.matchBuf = make(uint64Slice, len(r.prs))
+	}
+	mis := r.matchBuf[:len(r.prs)]
+	idx := 0
+	sorted := true
 	for _, p := range r.prs {
```
Comment (xiang90) [inline raft/raft.go]: it seems that it is low chance that prs happens to be sorted? not sure if this optimization makes sense.

Comment (siddontang) [reply raft/raft.go]: In TiKV, we still sort it every time. maybe we can write a benchmark for this.

Comment (nvanbenschoten) [reply raft/raft.go]: Yeah the change that this is sorted is low - `1/(n!)` if we assume that map iteration is always random. Probably not worth the extra complexity to this function. Removed.

File: raft/raft.go (hunk 1 above)
Comment (nvanbenschoten) [reply raft/raft.go]: Done.

**Comment by nvanbenschoten** [comment]
Reviving this after seeing this allocation account for `.92%` of allocations in a production CockroachDB cluster.

**Comment by xiang90** [comment]
lgtm. defer to @siddontang or @gyuho

**Comment by gyuho** [comment]
I am on vacation till Friday. Will review by Monday. Thanks!

**Pull request merged**